dizoo/metadrive/env/drive_env.py

```diff
@@ -0,0 +1,384 @@
+        long_last, _ = current_lane.local_coordinates(vehicle.last_position)
+        long_now, lateral_now = current_lane.local_coordinates(vehicle.position)
+
+        # reward for lane keeping, without it vehicle can learn to overtake but fail to keep in lane
+        if self.config["use_lateral_reward"]:
+            lateral_factor = clip(1 - 2 * abs(lateral_now) / vehicle.navigation.get_current_lane_width(), 0.0, 1.0)
+        else:
+            lateral_factor = 1.0
+
+        reward = 0.0
+        reward += self.config["driving_reward"] * (long_now - long_last) * lateral_factor * positive_road
+        reward += self.config["speed_reward"] * (vehicle.speed / vehicle.max_speed) * positive_road
+
+        step_info["step_reward"] = reward
+
+        if self._is_arrive_destination(vehicle):
+            reward = +self.config["success_reward"]
+        elif self._is_out_of_road(vehicle):
+            reward = -self.config["out_of_road_penalty"]
+        elif vehicle.crash_vehicle:
+            reward = -self.config["crash_vehicle_penalty"]
+        elif vehicle.crash_object:
+            reward = -self.config["crash_object_penalty"]
+        return reward, step_info
+
+    def _get_reset_return(self):
+        ret = {}
+        self.engine.after_step()
+        for v_id, v in self.vehicles.items():
+            self.observations[v_id].reset(self, v)
+            ret[v_id] = self.observations[v_id].observe(v)
+        return ret if self.is_multi_agent else self._wrap_as_single_agent(ret)
+
+    def switch_to_third_person_view(self) -> (str, BaseVehicle):
+        if self.main_camera is None:
+            return
+        self.main_camera.reset()
+        if self.config["prefer_track_agent"] is not None and self.config["prefer_track_agent"] in self.vehicles.keys():
+            new_v = self.vehicles[self.config["prefer_track_agent"]]
+            current_track_vehicle = new_v
+        else:
+            if self.main_camera.is_bird_view_camera():
+                current_track_vehicle = self.current_track_vehicle
+            else:
+                vehicles = list(self.engine.agents.values())
+                if len(vehicles) <= 1:
+                    return
+                if self.current_track_vehicle in vehicles:
+                    vehicles.remove(self.current_track_vehicle)
+                new_v = get_np_random().choice(vehicles)
+                current_track_vehicle = new_v
+        self.main_camera.track(current_track_vehicle)
+        return
+
+    def switch_to_top_down_view(self):
+        self.main_camera.stop_track()
+
+    def setup_engine(self):
+        super(MetaDrivePPOOriginEnv, self).setup_engine()
+        self.engine.accept("b", self.switch_to_top_down_view)
+        self.engine.accept("q", self.switch_to_third_person_view)
+        from metadrive.manager.traffic_manager import TrafficManager
+        from metadrive.manager.map_manager import MapManager
```

Comment:
why import here

Reply:
This setup_engine function is the official implementation of metadrive.